pkg/vault/contracts/BaseRouter.sol

```diff
@@ -0,0 +1,726 @@
+                _weth.unwrapWethAndTransferToSender(_vault, params.sender, amountOut);
+            } else {
+                // Transfer the token to the sender (amountOut).
+                _vault.sendTo(token, params.sender, amountOut);
+            }
+        }
+
+        _returnEth(params.sender);
+    }
+
+    /**
+     * @notice Hook for removing liquidity in Recovery Mode.
+     * @dev Can only be called by the Vault, when the pool is in Recovery Mode.
+     * @param pool Address of the liquidity pool
+     * @param sender Account originating the remove liquidity operation
+     * @param exactBptAmountIn BPT amount burned for the output tokens
+     * @param minAmountsOut Minimum amounts of tokens to be received, sorted in token registration order
+     * @return amountsOut Actual token amounts transferred in exchange for the BPT
+     */
+    function removeLiquidityRecoveryHook(
+        address pool,
+        address sender,
+        uint256 exactBptAmountIn,
+        uint256[] memory minAmountsOut
+    ) external nonReentrant onlyVault returns (uint256[] memory amountsOut) {
+        amountsOut = _vault.removeLiquidityRecovery(pool, sender, exactBptAmountIn, minAmountsOut);
+
+        IERC20[] memory tokens = _vault.getPoolTokens(pool);
+
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            uint256 amountOut = amountsOut[i];
+            if (amountOut > 0) {
+                // Transfer the token to the sender (amountOut).
+                _vault.sendTo(tokens[i], sender, amountOut);
+            }
+        }
+
+        _returnEth(sender);
+    }
+
+    /**
+     * @notice Hook for swaps.
+     * @dev Can only be called by the Vault. Also handles native ETH.
+     * @param params Swap parameters (see IRouter for struct definition)
+     * @return amountCalculated Token amount calculated by the pool math (e.g., amountOut for an exact in swap)
+     */
+    function swapSingleTokenHook(
+        SwapSingleTokenHookParams calldata params
+    ) external nonReentrant onlyVault returns (uint256) {
+        uint256 tokenInCredit;
+        if (_isAggregator) {
+            // `amountInHint` represents the amount supposedly paid upfront by the sender.
+            uint256 amountInHint;
+            if (params.kind == SwapKind.EXACT_IN) {
+                amountInHint = params.amountGiven;
+            } else {
+                amountInHint = params.limit;
+            }
+
+            // Always settle the amount paid first to prevent potential underflows at the vault. `tokenInCredit`
+            // represents the amount actually paid by the sender, which can be at most `amountInHint`.
+            // If the user paid less than what was expected, revert early.
+            tokenInCredit = _vault.settle(params.tokenIn, amountInHint);
+            if (tokenInCredit < amountInHint) {
+                revert InsufficientPayment(params.tokenIn);
+            }
+        }
```

Comment:
Following the comments above, I don't think settling before `swap` is really needed anymore. We should be able to just have a single branch checking `isAggregator` after calling `_swapHook`.